bridge: setup ebtables dedup rules if promisc mode is specified.

dnardo opened the pull request:
This is required to prevent duplicate packets when promiscuous mode is set on the bridge interface.
This is related to moving kubernetes "kubenet" functionality out of kubernetes and into the bridge plugin.
https://github.com/kubernetes/kubernetes/issues/38890

Original issue requiring the deduplication: https://github.com/kubernetes/kubernetes/issues/25793
Original Kubernetes PR fixing the issue: https://github.com/kubernetes/kubernetes/pull/28717

dcbw commented on pkg/bridge/bridge.go:
```diff
@@ -0,0 +1,89 @@
+		}
+	}
+
+	common_args := []string{"-p", "IPv4", "-s", mac, "-o", "veth+", "--ip-src"}
```
Hmm, how about IPv6, should we do the same there?

dcbw commented on pkg/bridge/ebtables/ebtables.go:
```diff
@@ -0,0 +1,180 @@
+	path string
+}
+
+func New() (*EBTables, error) {
```
Could we pass an 'execer' interface here instead of just using os/exec directly, so that we can mock the whole thing rather than using live ebtables on the host?

dcbw commented on pkg/bridge/ebtables/ebtables_test.go:
```diff
@@ -0,0 +1,95 @@
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("EBTables Operations", func() {
```
Might want to consider running these in their own network namespace so we don't pollute the host's?

squeed commented:
Might have to add another deb to the .travis.yml file.

dnardo commented:
@dcbw I think I've address your comments.  Added IPv6 support, and updated the test cases.  I added a lightweight os.exec interface and the test should now run in their own NS.
I wasn't sure if there some particular tests that you wanted mocked out, since the tests I added are mostly e2e/integration tests.

bboreham commented on pkg/exec/exec.go:
```diff
@@ -0,0 +1,66 @@
+	osexec "os/exec"
+)
+
+type cmdWrapper osexec.Cmd
```
Why do we need this wrapper?  (I.e. the entire file)

dnardo replied:
It was requested that it be wrapped by dcbw, I imagine for testing purposes.

squeed commented:
I know this issue isn't exactly the right place to discuss it, but is this still needed? From what I can gather from kubernetes/kubernetes#20096, this issue is still sometimes happening even with hairpin turned off, and it's not really clear if hairpin was the trigger.

I would love to get rid of the promisc + ebtables fix - they introduce a ton of code complexity. But, if it's needed, it's needed.

dnardo commented:
@squeed 
I agree with what you are saying.  We don't know that it was the cause, we do know that for the most part mitigated the issue although.  I'm hesitant to add this code in CNI as well, which is why I've not been pushing hard on this PR.  The task that was given to me was to deprecate "kubenet" and migrate existing bridge plugin work-arounds into the CNI bridge plugin.  That said one direction we could go with, is to just make p2p the default for Kubernetes plugin and don't put the work-arounds into CNI bridge plugin (and remove the already merged promisc stuff).  I'm not, however, comfortable with keeping the bridge plugin the default and not having the work arounds since we should have parity with what is part of kubenet today.

dnardo commented:
@thockin for his thoughts on this as well.

squeed commented:
@dnardo awesome, glad to see we're on the same page. However, I do suspect we'll have to add it, given that we already have released a version of bridge with promiscuous. Here's hoping Tim has the magic answer.

dnardo commented:
@thockin @squeed 
I think at the next sig-network for k8s I'll just float the idea of changing the default plugin to ptp.  That may be the quickest way to get this moving forward.

rpothier commented on pkg/bridge/ebtables/ebtables.go:
```diff
@@ -0,0 +1,180 @@
+)
+
+const (
+	sysctlBridgeCallIPTables = "net/bridge/bridge-nf-call-iptables"
```
Can we also set the v6 iptables?
sysctlBridgeCallIP6Tables = "net/bridge/bridge-nf-call-ip6tables"

rpothier commented on pkg/bridge/ebtables/ebtables.go:
```diff
@@ -0,0 +1,180 @@
+
+	// Allow iptables to see bridged traffic, do this best effort.
+	sysctl.Sysctl(sysctlBridgeCallIPTables, "1")
+
```
here also

rpothier replied:
Also, just an implementation question, why is this set here in the plugin and not in the proxier that actually uses it?

m1093782566 commented:
@dnardo @squeed 

From my test, CNI ptp plugin does not resolve the hairpin issue. In other words, container A -> VIP -> container A is not reachable in ptp model by default. 

So, if ptp plugin does no help to hairpin issue, why we don't support veth(not promisc-bridge mode) type hairpin for bridge plugin?

m1093782566 commented:
/cc @leblancd

squeed commented:
@m1093782566 CNI ptp definitely *should* support hairpin. Can you open an issue with a test case?

I've successfully done some hairpin traffic in connection with hostport mapping, so this may be a bad interaction with kube-proxy.

m1093782566 commented:
@squeed , I created an issue: #139.
